File: fibonacci/fibonacci.R
rm(list = ls())
# Fibonacci sequence up to n-th term using recursive functions
fibonacci <- data.frame(n = c(0, 1), result = c(1, 1))
str(fibonacci)
recursive.fibonacci <- function(n) {
  if(n <= 1) {
    return(n)
  } else if (n %in% fibonacci$n) {
    # if already computed
    row_index <- which(fibonacci$n == n)
    return (fibonacci %>% select(result) %>% dplyr::filter(row_number() == row_index))
  } else {
    # save the result and return it
    new_result <- recursive.fibonacci(n-1) + recursive.fibonacci(n-2)
    fibonacci %>% add_row(n = n, result = new_result)
    fibonacci[nrow(fibonacci) + 1,] = c(n, new_result)
    return(new_result)
  }
}

recursive.fibonacci(3)
recursive.fibonacci(5)

File: fibonacci/fibonacci.py
from time import time


class Fibonacci:
    def __init__(
            self,
            fibonacci = {0: 0, 1: 1}
        ) -> None:
        self.fibonacci = fibonacci
    
    def compute(self, n:int):
        '''
        Return
        ------
        int
        '''
        if n in self.fibonacci.keys():
            return self.fibonacci[n]
        # compute the Fibonacci number and save it in the dictionnary
        self.fibonacci[n] = self.compute(n - 1) + self.compute(n - 2)
        return self.fibonacci[n]

    def output_result(self, n:int):
        return print(f"The result of the Fibonacci sequence for n={n} is {self.compute(n=n)}")
    
    def __str__(self) -> str:
        return f"Fibonacci sequence\n{self.fibonacci}"


if __name__ == "__main__":
    fibonacci = Fibonacci()
    fibonacci.output_result(n=10)
    fibonacci.output_result(n=15)
    print(fibonacci)
